### PR TITLE
Make index and delete operation execute as a single bulk item

### PR DESCRIPTION
**Pull request by areek**
Performance testing by @danielmitterdorfer revealed single
index/delete operations have similar performance (indexing
throughput) to equivalent single item bulk request.
This PR reduces the code paths to executing single write
operations, by reusing the logic in (shard) bulk action for
executing single operation as a single-item bulk request.

### CODE REVIEW

**Comment by bleskes** [comment]
@areek thanks. I quickly looked at it and I have a question. What happens when people issue an index request now via the transport client? I believe they still go through the the transportIndexAction mechanics? we need to make sure it's translated and we never send a reroute request to the primaries (I expected some change in the execute method of that action). Also I think we need to change the way the rest layer work (we can do it as a follow up if you prefer)

**Comment by areek** [comment]
Thanks for the feedback @bleskes! 

> What happens when people issue an index request now via the transport client? I believe they still go through the the transportIndexAction mechanics?

when a transport client issues an index request it still goes through transportIndexAction (just like it does for node clients). The only difference in this PR is we use shardBulk action to execute the actual operation. 

>  we need to make sure it's translated and we never send a reroute request to the primaries (I expected some change in the execute method of that action)

I took this approach initially, i.e. delegate to bulk action on the transportIndexAction's `doExecute`. But tests started failing as bulk action is not a replicated action and as a result behaves differently (e.g. handling cluster state global blocks, in bulk action we get an exception outright ([TransportBulkAction#209](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java#L209)) whereas for replication action we retry on global blocks ([TransportReplicationAction#737](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java#L737)). Should we change bulk action to have similar behaviour for 'reroute' phase?

**Comment by areek** [comment]
@bleskes I have updated the PR to use single-item bulk action to execute index and delete operation. Now we delegate single write operation requests to bulk action in `doExecute` method, this simplified `TransportIndexAction` & `TransportDeleteAction` s to `HandledTransportAction`. There are a few functional changes in the PR, namely:
 - `TransportBulkAction` now retries retryable cluster blocks (to be consistent with `TransportReplicationAction`)
 - The `toString` representation of `IndexRequest` and `DeleteRequest` has changed to be the same as a single `ShardBulkRequest.toString`, should we special case this? This might lead to confusion while debugging
 - Temporarily I increased the threadpool size for `Bulk` to make tests pass from `50` to `200` (default size for index threadpool). As a side, now only update action uses the index threadpool
 - All tests pass except `IndexWithShadowReplicasIT`, am still investigating the cause.

**Comment by martijnvg** [comment]
@areek I think with this change, we don't need the logic for IndexRequests in `IngestActionFilter`? (switch case for index action in apply() method, processIndexRequest() method and PipelineExecutionService#executeIndexRequest(...) method). If it isn't that much work, maybe we can remove it as part of this change?

File: core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 1]
```diff
@@ -169,7 +169,7 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
         final int genericThreadPoolMax = boundedBy(4 * availableProcessors, 128, 512);
         builders.put(Names.GENERIC, new ScalingExecutorBuilder(Names.GENERIC, 4, genericThreadPoolMax, TimeValue.timeValueSeconds(30)));
         builders.put(Names.INDEX, new FixedExecutorBuilder(settings, Names.INDEX, availableProcessors, 200));
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Is there a reason to keep `INDEX` TP around?

Comment (areek) [inline core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: `INDEX` TP is currently only used by `TransportUpdateAction`. I agree that we should remove it or at least rename it to better reflect where it is used.

**Comment by areek** [comment]
thanks @martijnvg for the comment, I will remove the logic index requests logic in `IngestActionFilter`

File: core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 1 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: I see, right, renaming it to update would be better.

**Comment by areek** [comment]
@bleskes Can you take a look, I reverted index and delete transport actions to be `TransportWriteAction`s and now it delegates the `shardOperationOnPrimary/Replica` to reuse the shard bulk logic for performing single-item shard bulks. This was done to ensure bwc, such that incoming primary/replica single-operation-write-requests from pre-6.0 nodes are executed as expected, while index/delete requests from 6.0 nodes are delegated to TransportBulk and avoids the replication reroute phase entirely.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 2]
```diff
@@ -420,4 +425,113 @@ protected WriteReplicaResult shardOperationOnReplica(BulkShardRequest request, I
         return next;
     }
 
+    public <Request extends ReplicatedWriteRequest<Request>, Response extends ReplicationResponse & WriteResponse>
+    WritePrimaryResult<Request, Response> executeSingleItemBulkRequestOnPrimary(
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: wondering if we can put this something else .. maybe a static method on TransportIndexAction ? It's a shame to cluster this class with BWC code..

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 3]
```diff
@@ -420,4 +425,113 @@ protected WriteReplicaResult shardOperationOnReplica(BulkShardRequest request, I
...
+            Request request, IndexShard primary) throws Exception {
+        BulkItemRequest[] itemRequests = new BulkItemRequest[1];
+        WriteRequest.RefreshPolicy refreshPolicy = request.getRefreshPolicy();
+        request.setRefreshPolicy(WriteRequest.RefreshPolicy.NONE);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: feels weird to mutate the incoming request here. Why is it needed?

File: core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java [hunk 4]
```diff
@@ -48,72 +42,48 @@
 /**
  * Performs the delete operation.
  */
-public class TransportDeleteAction extends TransportWriteAction<DeleteRequest, DeleteRequest,DeleteResponse> {
+public class TransportDeleteAction extends TransportWriteAction<DeleteRequest, DeleteRequest, DeleteResponse> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java]: wondering - should we mark this as deprecated, so we want use internally? not sure how much it buys us.

File: core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java [hunk 5]
```diff
@@ -48,72 +42,48 @@
...
+                                 TransportBulkAction bulkAction, TransportShardBulkAction shardBulkAction) {
+        super(settings, DeleteAction.NAME, transportService, clusterService, indicesService, threadPool, shardStateAction,
+                actionFilters, indexNameExpressionResolver, DeleteRequest::new, DeleteRequest::new, ThreadPool.Names.INDEX);
+        this.bulkAction = bulkAction;
+        this.shardBulkAction = shardBulkAction;
     }
 
     @Override
     protected void doExecute(Task task, final DeleteRequest request, final ActionListener<DeleteResponse> listener) {
-        ClusterState state = clusterService.state();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java]: Since we can't override the execute method itself, I think it will be good to get a sanity check from @jaymode and @martijnvg about potential implications of running the request filter chain twice

File: core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java [hunk 6]
```diff
@@ -67,93 +51,47 @@
...
+        BulkRequest bulkRequest = new BulkRequest();
+        bulkRequest.add(request);
+        bulkRequest.setRefreshPolicy(request.getRefreshPolicy());
+        bulkRequest.timeout(request.timeout());
+        bulkRequest.waitForActiveShards(request.waitForActiveShards());
+        request.setRefreshPolicy(WriteRequest.RefreshPolicy.NONE);
+        bulkAction.execute(task, bulkRequest, new ActionListener<BulkResponse>() {
+            @Override
+            public void onResponse(BulkResponse bulkItemResponses) {
+                assert bulkItemResponses.getItems().length == 1: "expected only one item in bulk request";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java]: since this is also the same code here as in delete, I wonder if we should add a generic BWC class that will do all this shared munging (and the code from the bulk action as well).

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java [hunk 7]
```diff
@@ -77,19 +78,24 @@ protected TransportWriteAction(Settings settings, String actionName, TransportSe
...
     @Override
-    protected abstract WriteReplicaResult shardOperationOnReplica(ReplicaRequest request, IndexShard replica) throws Exception;
+    protected abstract WriteReplicaResult<ReplicaRequest> shardOperationOnReplica(
+            ReplicaRequest request, IndexShard replica) throws Exception;
 
     /**
      * Result of taking the action on the primary.
      */
-    protected class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult {
+    protected static class WritePrimaryResult<ReplicaRequest extends ReplicatedWriteRequest<ReplicaRequest>,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java]: wondering - why was this needed (the change to static and the explicit generic references)?

File: core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: I think we can migrate `TransportUpdateAction` to redirect to a single item bulk as well (as a follow up). At that point we can rename bulk to index :)

File: core/src/test/java/org/elasticsearch/action/index/TransportIndexActionIngestTests.java [hunk 8]
```diff
@@ -1,224 +0,0 @@
...
-import static org.mockito.Matchers.same;
-import static org.mockito.Mockito.doAnswer;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.reset;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.verifyZeroInteractions;
-import static org.mockito.Mockito.when;
-
-public class TransportIndexActionIngestTests extends ESTestCase {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/index/TransportIndexActionIngestTests.java]: do we have any test that makes sure that single item index ops work with ingest?

File: core/src/test/java/org/elasticsearch/cluster/NoMasterNodeIT.java [hunk 9]
```diff
@@ -121,23 +121,21 @@ public void run() {
...
                 client().prepareUpdate("no_index", "type1", "1")
                         .setScript(new Script(
                             ScriptType.INLINE, Script.DEFAULT_SCRIPT_LANG, "test script", Collections.emptyMap())).setTimeout(timeout));
 
 
-        checkWriteAction(false, timeout,
+        checkWriteAction(timeout,
                 client().prepareIndex("test", "type1", "1").setSource(XContentFactory.jsonBuilder().startObject().endObject()).setTimeout(timeout));
 
-        checkWriteAction(autoCreateIndex, timeout,
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/NoMasterNodeIT.java]: why can we remove autoCreateIndex? I mean it can still be created? what am I missing?

File: core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java [hunk 10]
```diff
@@ -57,14 +60,8 @@
...
     private ClusterService clusterService;
-    private Transport transport;
     private TransportService transportService;
-    private IndicesService indicesService;
-    private ShardStateAction shardStateAction;
-    private ActionFilters actionFilters;
-    private IndexNameExpressionResolver indexNameExpressionResolver;
-    private AutoCreateIndex autoCreateIndex;
-    private Settings settings;
+    private TransportIndexAction transportIndexAction;
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java]: we should stop using it :) - maybe have a test utility to index a single item using a bulk action?

**Review by bleskes — CHANGES_REQUESTED** [review]
This looks great. Left very minor comments and questions. I also asked @jaymode and @martijnvg for some validation.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 2 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I have moved these BWC method to static methods to `SingleWriteOperationUtility`, (suggestion for better name welcomed). The only downside is now `shardOperationOnPrimary` and `shardOperationOnReplica` for bulk shard action is made public, as they are called from transport index and delete actions

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 3 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: thanks for pointing it out. This was left over from using the bulk action directly, I removed mutating the request

File: core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java (hunk 4 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java]: I marked both transport index/delete action as deprecated, now the only other usage of these actions are in TransportUpdateAction. should we use the bulk action directly as a follow up?

File: core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java (hunk 6 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java]: I converted them to static helper functions and moved them to `SingleWriteOperationUtilty`

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java (hunk 7 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java]: this is needed because we deal with two types of `WritePrimaryResult` in `indexAction/deleteAction.shardOperationOnPrimary/Replica`. `WritePrimaryResult` with type `BulkShardRequest/BulkShardResponse` and `Index`/`delete` request and response.

File: core/src/test/java/org/elasticsearch/action/index/TransportIndexActionIngestTests.java (hunk 8 above)
Comment (areek) [inline core/src/test/java/org/elasticsearch/action/index/TransportIndexActionIngestTests.java]: There is already test for ingest with bulk action (`TransportBulkActionIngestTests`) which covers ingest with single item ops

File: core/src/test/java/org/elasticsearch/cluster/NoMasterNodeIT.java (hunk 9 above)
Comment (areek) [inline core/src/test/java/org/elasticsearch/cluster/NoMasterNodeIT.java]: This is actually due to `TransportUpdateAction` behaving differently (on how it handles cluster blocks when `autoCreateIndex` is `true`, instead of throwing cluster block exception, transport update throws master not discovered exception). IMO, update action should be fixed to be consistent here and throw a cluster block exception instead. I re-worked the tests, so the distinction is clear.

**Comment by areek** [comment]
Thanks @bleskes for the feedback, I addressed your comments. After running bwc tests `backward-5.0` for many iterations, I found a bug where the listener hangs indefinitely when the refresh flag is set to `WAIT_FOR` (one of rest-tests testing refresh for bulk, index and delete seemed to fail). I couldn't reproduce this running the rest-test on only 6.0 cluster. I am investigating the code flow in `TransportWriteAction.AsyncAfterWriteAction` but still no luck.

**Comment by bleskes** [comment]
thx @areek . I played with this a bit and came to this: https://github.com/areek/elasticsearch/compare/enhancement/use_shard_bulk_for_single_ops...bleskes:use_shard_bulk_for_single_ops

What do you think?

Also - I think the wait for refresh flag issues are unrelated. They have been showing up on normal CI. I started chasing them too

File: core/src/main/java/org/elasticsearch/action/bulk/TransportSingleItemBulkWriteAction.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    @Override
+    protected WriteReplicaResult<Request> shardOperationOnReplica(
+        Request replicaRequest, IndexShard replica) throws Exception {
+        BulkItemRequest[] itemRequests = new BulkItemRequest[1];
+        WriteRequest.RefreshPolicy refreshPolicy = replicaRequest.getRefreshPolicy();
+        itemRequests[0] = new BulkItemRequest(0, ((DocWriteRequest) replicaRequest));
+        BulkShardRequest bulkShardRequest = new BulkShardRequest(replicaRequest.shardId(), refreshPolicy, itemRequests);
+        WriteReplicaResult<BulkShardRequest> result = shardBulkAction.shardOperationOnReplica(bulkShardRequest, replica);
+        // nocommit - is the null failure ok?
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportSingleItemBulkWriteAction.java]: @bleskes currently, a replica operation can never throw a document-level failure, as the same document has been already indexed successfully in the primary, so passing a null failure here is okay. all other failures (e.g. persistent shard failures) will bubble up

**Comment by areek** [comment]
> What do you think?

@bleskes I like it, I cherry-picked your commit and made it a part of the PR and left a comment on your no-commit comment.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportSingleItemBulkWriteAction.java (hunk 11 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportSingleItemBulkWriteAction.java]: OK. Let's remove the commit then.

**Comment by bleskes** [comment]
retest this please

**Review by bleskes — APPROVED** [review]
LGTM. @martijnvg do you mind taking a look at the ingest tests?

**Comment by bleskes** [comment]
@danielmitterdorfer has gracefully agreed to benchmark this PR to validate that performance is comparable to his initial research

File: core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java [hunk 12]
```diff
@@ -200,6 +235,38 @@ public void testIngestLocal() throws Exception {
...
+        verify(executionService).executeBulkRequest(bulkDocsItr.capture(), failureHandler.capture(), completionHandler.capture());
+        completionHandler.getValue().accept(exception);
+        assertTrue(failureCalled.get());
+
+        // now check success
+        indexRequest.setPipeline(null); // this is done by the real pipeline execution service when processing
+        completionHandler.getValue().accept(null);
+        assertTrue(action.isExecuted);
+        assertFalse(responseCalled.get()); // listener would only be called by real index action, not our mocked one
+        verifyZeroInteractions(transportService);
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java]: I think we should also verify that interaction happened with `ingestService` mock here?

**Review by martijnvg — APPROVED** [review]
Looks good, left one minor comment.

File: core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java (hunk 12 above)
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java]: @bleskes:

```patch
Index: core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java	(date 1482389279000)
+++ core/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java	(revision )
@@ -28,15 +28,12 @@
 import org.elasticsearch.cluster.ClusterChangedEvent;
 import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.ClusterStateApplier;
-import org.elasticsearch.cluster.action.shard.ShardStateAction;
-import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.node.DiscoveryNodes;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.collect.ImmutableOpenMap;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.concurrent.AtomicArray;
-import org.elasticsearch.indices.IndicesService;
 import org.elasticsearch.ingest.IngestService;
 import org.elasticsearch.ingest.PipelineExecutionService;
 import org.elasticsearch.tasks.Task;
@@ -54,7 +51,6 @@
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
-import java.util.function.Supplier;
 
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.sameInstance;
@@ -64,6 +60,7 @@
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyZeroInteractions;
 import static org.mockito.Mockito.when;
@@ -224,6 +221,7 @@
         verify(executionService).executeBulkRequest(bulkDocsItr.capture(), failureHandler.capture(), completionHandler.capture());
         completionHandler.getValue().accept(exception);
         assertTrue(failureCalled.get());
+        verify(executionService, times(1)).executeBulkRequest(any(), any(), any());
 
         // now check success
         Iterator<DocWriteRequest> req = bulkDocsItr.getValue().iterator();
@@ -311,6 +309,7 @@
         } else {
             assertSame(remoteNode1, node.getValue());
         }
+        verifyZeroInteractions(executionService);
     }
 
     public void testSingleItemBulkActionIngestForward() throws Exception {

```

**Comment by danielmitterdorfer** [comment]
@areek, @bleskes I have benchmarked the single index operation with async translog fsync *) of the latest version on Areek's branch ("contender") against the merge base (baseline 8aca504c863d2351258e33e86b2c18e9365bbf99). The overhead is basically not measurable:

| Metric            | Baseline Value | Contender Value | Unit   | 
|-------------------|----------------|-----------------|--------| 
| Min Throughput    | 9410           | 9345            | docs/s | 
| Median Throughput | 9439           | 9433            | docs/s | 
| Max Throughput    | 9454           | 9447            | docs/s | 

I used a custom track ([ltaxis](https://github.com/danielmitterdorfer/ct/tree/master/ltaxis)) which is a stripped down version of the standard nyc_taxis track ("only" 15 million records). The main reason is that this benchmark is doing so much requests that it would not be feasible to run the full nyc_taxis track.

_*) async translog fsync is enabled to avoid accidentally bottlenecking on the fsync operation._
